src/channel-qvm.lisp

```diff
@@ -0,0 +1,374 @@
+  "Does the noisy qvm QVM require swapping of internal pointers?"
+  (and (not (eq (amplitudes qvm) (original-amplitude-pointer qvm)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod run :after ((qvm channel-qvm))
+  ;; Only copy if we really need to.
+  (when (requires-swapping-p qvm)
+    ;; Copy the correct amplitudes into place.
+    (copy-wavefunction (amplitudes qvm) (original-amplitude-pointer qvm))
+    ;; Get the pointer back in home position. We want to swap them,
+    ;; not overwrite, because we want the scratch memory to be intact.
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))))
+
+
+(defmethod set-readout-povm ((qvm channel-qvm) qubit povm)
+  " Set a readout povm for a qubit on this qvm"
+  (check-povm povm)
+  (setf (gethash qubit (readout-povms (noise-model qvm))) povm)
+  nil)
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
+  "If any noise rules are matched by this instruction, apply the the kraus operators
+for that noise rule after applying the gate for this instruction
+"
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar #'(lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (instr-qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (instr-name (list (cl-quil::application-operator-name instr)))
+         (noise-rules (noise-rules (noise-model qvm)))
+         (prepend-kraus-ops (match-rules noise-rules instr-qubits instr-name "before"))
+         (append-kraus-ops (match-rules noise-rules instr-qubits instr-name "after"))
+         )
+    (when prepend-kraus-ops (apply-all-kraus-operators qvm instr prepend-kraus-ops))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple instr-qubits) params)
+    (when append-kraus-ops (apply-all-kraus-operators qvm instr append-kraus-ops))
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod transition :around ((qvm channel-qvm) (instr quil:measurement))
+  (let ((ret-qvm (call-next-method)))
+    (apply-classical-readout-noisee ret-qvm instr)
+    ret-qvm))
+
+
+(defun matches-rule (rule instr-qubits instr-name position)
+  "check if rule is matched by instruction data and the position of the match request"
+  (let* ((noise-predicate (predicate (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr-qubits instr-name) (string= position noise-position))))
+
+
+(defun match-rules (rules instr-qubits instr-name position)
+  "return the operation elements for the first matching rule"
+  (loop :for rule in rules
+        :when (matches-rule rule instr-qubits instr-name position)
+          :return (operation-elements rule)))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-operators)
```

Comment:
Not strictly required, but I *think* (could be wrong) we follow the convention of always explicitly specifying an explicit `defgeneric` for any methods that we add. Search, e.g. `defgeneric run` in `src/qam.lisp`.